### DISCUSSION TITLE
How is the Machine ID created when a new machine is added?

### DISCUSSION THREAD

**Discussion by danross826** [cmu-sei/GHOSTS#336] [General]
So, I have my ghosts api set up. Whenever I install a ghosts client on a new machine and link to my ghosts api server, it adds the machine to the database with a new machine id. However, I want to use the api to add machines before I install the client on them. Every machine has a machine id that's created an assigned to it when the client connects. How is this machine Id created? I've been looking through the code and I can see a function called at various points that creates a GUID. Is that function in the code somewhere (I can't find it if it is) or is it an inbuilt function of the language ghosts is using?

**Comment by sei-dupdyke** [comment]
This is a use case I had never planned on. I suspect you could fill in your machine entry with whatever the client will eventually pass to the server and it would send back the ID for the record you've created. But if any value does not match up, then the server will create a new record.

What happens now is that:

1. The client makes a request, if it has already gotten an ID from the server, this is stored in instance/id.json
2. The server looks the client up, if there is an ID, this is simple, if not, it is done by default via:

machine.Name
machine.FQDN.ToLower()
machine.Host.ToLower()
machine.ResolvedHost.ToLower()

You can also configure the server to check by a specific value as well.

Why do you want to create the record first? Just curious please.

**Reply by danross826** [reply]
I want to create timelines for machines that don't exist yet basically. They want to create a range of VM's that will have the ghosts client installed on them automatically and then the machine will connect with the API server and be connected with the machine record and timelines connected to it. To do this, I'm taking the values that will be used from a Terraform script to create the VM's and turning them into JSON so that I can upload them as machines to the API server. 

My hurdles to this so far seem to be that it doesn't accept the Machine ID that I add to the machine record. I've tried both leaving it blank as well as creating a method that assigns it a string random characters separated by hyphens so that it appears as a natural machine id that would be assigned. The other hurdle that I haven't really explored yet is that it also doesn't like the time stamp for the machine record. I'm guessing there's something off about my formatting of it but haven't explored that yet.

**Reply by sei-dupdyke** [reply]
Seems reasonable. I am going to turn this into an issue will look into this when I have a few cycles.

I think this also fits in well to generating full animator NPC profiles in the future...

**Reply by sei-dupdyke** [reply]
#337
